src/win/i2c.cc

```diff
@@ -0,0 +1,268 @@
+#include "i2c.h"
+#include <ppltasks.h>
+#include <vector>
+#include <iostream>
+#include "..\util.h"
+
+using v8::FunctionTemplate;
+using namespace concurrency;
+using namespace Platform;
+using namespace Windows::Devices::Enumeration;
+
+Nan::Persistent<v8::Function> WinI2c::constructor;
+
+NAN_MODULE_INIT(WinI2c::Init) {
+  v8::Local<v8::FunctionTemplate> tpl = Nan::New<v8::FunctionTemplate>(New);
+  tpl->SetClassName(Nan::New("WinI2c").ToLocalChecked());
+  tpl->InstanceTemplate()->SetInternalFieldCount(6);
+
+  Nan::SetPrototypeMethod(tpl, "getController", GetController);
+  Nan::SetPrototypeMethod(tpl, "getControllerSync", GetControllerSync);
+  Nan::SetPrototypeMethod(tpl, "closeDevice", CloseDevice);
+  Nan::SetPrototypeMethod(tpl, "closeDeviceSync", CloseDeviceSync);
+  Nan::SetPrototypeMethod(tpl, "readPartial", ReadPartial);
+  Nan::SetPrototypeMethod(tpl, "readPartialSync", ReadPartialSync);
+  Nan::SetPrototypeMethod(tpl, "writePartial", WritePartial);
+  Nan::SetPrototypeMethod(tpl, "writePartialSync", WritePartialSync);
+  Nan::SetPrototypeMethod(tpl, "createDevice", CreateDevice);
+  Nan::SetPrototypeMethod(tpl, "createDeviceSync", CreateDeviceSync);
+  Nan::SetPrototypeMethod(tpl, "writeReadPartial", WriteReadPartial);
+  Nan::SetPrototypeMethod(tpl, "writeReadPartialSync", WriteReadPartialSync);
+
+  constructor.Reset(Nan::GetFunction(tpl).ToLocalChecked());
+  Nan::Set(target, Nan::New("WinI2c").ToLocalChecked(), Nan::GetFunction(tpl).ToLocalChecked());
+}
+
+WinI2c::WinI2c() {
+  RoInitialize(RO_INIT_MULTITHREADED);
+}
+
+WinI2c::~WinI2c() {
+  RoUninitialize();
+}
+
+NAN_METHOD(WinI2c::New) {
+  if (info.IsConstructCall()) {
+    WinI2c *obj = new WinI2c();
+    obj->Wrap(info.This());
+    info.GetReturnValue().Set(info.This());
+  }
+  else {
+    const int argc = 1;
+    v8::Local<v8::Value> argv[argc] = { info[0] };
+    v8::Local<v8::Function> cons = Nan::New(constructor);
+    info.GetReturnValue().Set(cons->NewInstance(argc, argv));
+  }
+}
+
+class GetControllerWorker : public I2cAsyncWorker {
+public:
+    GetControllerWorker(
+    Nan::Callback *callback,
+    WinI2c* obj,
+    std::wstring controllerName
+  ) : I2cAsyncWorker(callback), obj(obj), controllerName(controllerName) { }
+
+  ~GetControllerWorker() {}
+
+  void Execute() {   
+    String^ deviceSelector = I2cDevice::GetDeviceSelector(ref new Platform::String(controllerName.c_str()));
+    DeviceInformationCollection^ i2cDeviceControllers = create_task(DeviceInformation::FindAllAsync(deviceSelector)).get();
+    
+    if (nullptr == i2cDeviceControllers) {
+      return Nan::ThrowError(Nan::ErrnoException(EPERM, "getController", "DeviceInformation::FindAllAsync failed to return controller(s)"));
```

Comment:
This isn't a good idea. The Execute method is not called on the event loop thread (the main thread.) It's called on a worker thread so it's important steer clear of everything related to NAN, uv, Node, and V8. The exception to this rule is uv_async_send().

The best thing to do at the moment is to replace `return Nan::ThrowError(...);` with
```cpp
SetErrorNo(EPERM);
SetErrorSyscall("getController");
return;
```
Unfortunately the nice error message gets lost. [I2cAsyncWorker](https://github.com/fivdi/i2c-bus/blob/master/src/util.h#L4-L64) could be extended to support the nice error message.